File: Intatis-Windows/src/Intatis.Shared/WorkspaceTools.cs
namespace Intatis.Windows.Shared;

public static class WorkspaceTools
{
    public static string ResolveWorkspace(string? configured, string? requested)
    {
        var source = !string.IsNullOrWhiteSpace(requested) ? requested : configured;
        if (string.IsNullOrWhiteSpace(source))
            throw new InvalidOperationException("No workspace path is configured.");

        var expanded = CommandParser.ExpandTilde(source);
        var full = Path.GetFullPath(expanded);
        if (!Directory.Exists(full))
            throw new DirectoryNotFoundException($"Workspace not found: {full}");
        return full;
    }

    public static IReadOnlyList<string> List(string workspace, string relativePath = "")
    {
        var dir = ResolvePath(workspace, relativePath);
        EnsureInsideWorkspace(workspace, dir);
        return Directory.EnumerateFileSystemEntries(dir)
            .Select(Path.GetFileName)
            .Where(n => !string.IsNullOrWhiteSpace(n))
            .OrderBy(n => n, StringComparer.OrdinalIgnoreCase)
            .Cast<string>()
            .ToList();
    }

    public static string ReadText(string workspace, string relativePath)
    {
        var path = ResolvePath(workspace, relativePath);
        EnsureInsideWorkspace(workspace, path);
        if (!File.Exists(path))
            throw new FileNotFoundException("File not found.", path);
        return File.ReadAllText(path);
    }

    public static IReadOnlyList<SearchHit> Search(string workspace, string needle, string? relativePath = null)
    {
        var root = ResolvePath(workspace, relativePath ?? ".");
        EnsureInsideWorkspace(workspace, root);

        if (!Directory.Exists(root) && !File.Exists(root))
            throw new FileNotFoundException("Path not found.", root);

        var matches = new List<SearchHit>();
        if (File.Exists(root))
        {
            SearchInFile(root, needle, matches, workspace);
            return matches;
        }

        foreach (var file in Directory.EnumerateFiles(root, "*.*", SearchOption.AllDirectories))
        {
            SearchInFile(file, needle, matches, workspace);
        }

        return matches;
    }

    public static void WriteText(string workspace, string relativePath, string content, bool overwrite = true)
    {
        var path = ResolvePath(workspace, relativePath);
        EnsureInsideWorkspace(workspace, path);
        var parent = Path.GetDirectoryName(path) ?? workspace;
        Directory.CreateDirectory(parent);
        if (!overwrite && File.Exists(path))
            throw new IOException("File already exists.");
        File.WriteAllText(path, content);
    }

    private static void SearchInFile(string path, string needle, List<SearchHit> matches, string workspace)
    {
        var lines = File.ReadLines(path);
        var relative = Path.GetRelativePath(workspace, path);
        var idx = 1;
        foreach (var line in lines)
        {
            if (line.Contains(needle, StringComparison.OrdinalIgnoreCase))
                matches.Add(new SearchHit(relative, idx, line));
            idx++;
        }
    }

    private static string ResolvePath(string workspace, string relativePath)
    {
        var fullWorkspace = Path.GetFullPath(workspace);
        if (string.IsNullOrWhiteSpace(relativePath) || relativePath == ".")
            return fullWorkspace;
        if (Path.IsPathRooted(relativePath))
            return Path.GetFullPath(relativePath);
        return Path.GetFullPath(Path.Combine(fullWorkspace, relativePath));
    }

    private static void EnsureInsideWorkspace(string workspace, string path)
    {
        var fullWorkspace = Path.GetFullPath(workspace)
            .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        var fullWorkspaceWithSlash = fullWorkspace + Path.DirectorySeparatorChar;
        var fullPath = Path.GetFullPath(path);
        if (!(string.Equals(fullPath, fullWorkspace, StringComparison.OrdinalIgnoreCase) ||
              fullPath.StartsWith(fullWorkspaceWithSlash, StringComparison.OrdinalIgnoreCase)))
        {
            throw new UnauthorizedAccessException("Access denied: path escapes workspace.");
        }
    }
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/WorkspaceTools.kt
package com.intatis.shared

import java.io.File

object WorkspaceTools {
    fun resolveWorkspace(configured: String?, requested: String?): String {
        val source = requested?.ifBlank { null } ?: configured?.ifBlank { null }
        if (source.isNullOrBlank()) {
            throw IllegalStateException("No workspace path is configured.")
        }

        val expanded = CommandParser.expandTilde(source)
        val full = File(expanded).absoluteFile
        if (!full.exists()) {
            throw IllegalArgumentException("Workspace not found: ${full.absolutePath}")
        }
        return full.absolutePath
    }

    fun list(workspace: String, relativePath: String = ""): List<String> {
        val root = resolvePath(workspace, relativePath)
        ensureInsideWorkspace(workspace, root)
        return root.list()
            ?.sorted()
            ?.toList()
            ?: emptyList()
    }

    fun readText(workspace: String, relativePath: String): String {
        val path = resolvePath(workspace, relativePath)
        ensureInsideWorkspace(workspace, path)
        if (!path.exists()) throw IllegalArgumentException("File not found: ${path.path}")
        return path.readText()
    }

    fun search(workspace: String, needle: String, relativePath: String? = null): List<SearchHit> {
        val root = resolvePath(workspace, relativePath ?: ".")
        ensureInsideWorkspace(workspace, root)

        if (!root.exists()) throw IllegalArgumentException("Path not found: ${root.path}")

        val matches = mutableListOf<SearchHit>()
        if (root.isFile) {
            searchInFile(root, workspace, needle, matches)
            return matches
        }

        root.walkTopDown()
            .onEnter { path -> path == root || !isHiddenPath(path) }
            .filter { it.isFile && !isHiddenPath(it) }
            .forEachIndexed { index, file ->
            if (matches.size >= 200) return matches
            searchInFile(file, workspace, needle, matches)
        }
        return matches
    }

    fun writeText(workspace: String, relativePath: String, content: String, overwrite: Boolean = true) {
        val path = resolvePath(workspace, relativePath)
        ensureInsideWorkspace(workspace, path)
        if (!overwrite && path.exists()) throw IllegalStateException("File already exists: ${path.path}")
        path.parentFile?.mkdirs()
        path.writeText(content)
    }

    private fun searchInFile(file: File, workspace: String, needle: String, matches: MutableList<SearchHit>) {
        val rel = file.relativeTo(File(workspace)).path
        var idx = 1
        file.forEachLine { line ->
            if (line.contains(needle, ignoreCase = false)) {
                matches.add(SearchHit(rel, idx, line))
            }
            idx++
            if (matches.size >= 200) return
        }
    }

    private fun resolvePath(workspace: String, relativePath: String): File {
        val root = File(workspace).absoluteFile
        return if (relativePath.isBlank() || relativePath == ".") {
            root
        } else if (File(relativePath).isAbsolute) {
            File(relativePath).absoluteFile
        } else {
            File(root, relativePath).absoluteFile
        }
    }

    private fun ensureInsideWorkspace(workspace: String, candidate: File) {
        if (!WorkspaceSecurity.isWithinWorkspace(candidate.absolutePath, File(workspace).absolutePath)) {
            throw IllegalArgumentException("Access denied: path escapes workspace")
        }
    }

    private fun isHiddenPath(file: File): Boolean {
        return file.isHidden || file.name.startsWith(".")
    }
}
